src/storage/src/render/persist_sink.rs

```diff
@@ -131,17 +163,32 @@ pub fn render<G>(
                     // per-timestamp.
                     for (row, ts, diff) in buffer.drain(..) {
                         if write.upper().less_equal(&ts) {
-                            stashed_batches
-                                .entry(ts)
-                                .or_insert_with(|| {
+                            let builder = stashed_batches.entry(ts).or_insert_with(|| {
+                                BatchBuilderAndCounts::new(write.builder(
                                     // TODO: the lower has to be the min because we don't know
                                     // what the minimum ts of data we will see is. In the future,
                                     // this lower should be declared in `finish` instead.
-                                    write.builder(100, Antichain::from_elem(Timestamp::minimum()))
-                                })
+                                    100,
+                                    Antichain::from_elem(Timestamp::minimum()),
+                                ))
+                            });
+
+                            let is_value = row.is_ok();
+                            builder
+                                .builder
                                 .add(&SourceData(row), &(), &ts, &diff)
                                 .await
                                 .expect("invalid usage");
+
+                            // Note that we assume `diff` is either +1 or -1 here, being anything
+                            // else is a logic bug we can't handle at the metric layer. We also
+                            // assume this addition doesn't overflow.
+                            match (is_value, diff.is_positive()) {
+                                (true, true) => builder.inserts += diff.unsigned_abs(),
```

Comment:
if we always take `unsigned_abs` and assume it's 1/-1, can we just do `counter += 1`?

Reply:
We assume them to be +1/-1 for the stats to make sense, but I want to leave them like this because they may very well help us debug a diff bug in the future!